rfcs/45-stageless.md

```diff
@@ -0,0 +1,725 @@
+    2. Fixed time-step pattern
+10. Port the entire engine to the new paradigm
+    1. We almost certainly need to port the improved ambiguity checker over to make this new paradigm usable
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until step 10.
+
+Lets explore some of the more critical and challenging details.
+
+### Storing and configuring systems
+
+Schedules store systems in a configured, initialized form.
+
+```rust
+struct Schedule{
+   // Each schedule is associated with exactly one `World`, as systems must be initiliazed on it
+   world_id: WorldId
+   // We need to be able to quickly look up specific systems and iterate over them
+   // `SystemId` should be a unique opaque identifier, generated on system insertion
+   // We cannot use a Vec + a `usize` index as the identifier,
+   // as the identifier would not be stable when systems were removed, risking serious bugs
+   systems: HashMap<SystemId, ConfiguredSystem>,
+   // Stores the fully initialized, efficient graph of ordering constraints
+   // for all systems in the schedule.
+   // The properties of each set are passed down to specific systems to cache computation results.
+   // `Graph` is a bikeshed-avoidance graph storage structure,
+   // the exact strategy should be benchmarked
+   ordering_constraints: Graph<SystemId>,
+   // Set configuration must be stored at the schedule level,
+   // rather than attached to specific set instances
+   sets: HashMap<Box<dyn SystemLabel>, SystemSetConfig>,
+   // Used to check if mutation is allowed
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: System,
+   config: SystemConfig,
+}
+```
+
+Raw systems store metadata, cached state, `Local` system parameters and the function pointer to the actual function to be executed:
+
+```rust
+struct System {
+   function: F,
+   // Contains all metadata and state
+   meta: SystemMeta
+   last_change_tick: u32,
+}
+```
+
+System configuration stores ordering dependencies, run criteria and system sets.
+It can be created raw, or stored in association with either system sets or systems.
+
+```rust
+enum OrderingTarget {
+   Label(Box<dyn SystemLabel>),
+   System(SystemId),
+}
+
+struct SystemConfig {
+   strict_ordering_before: Vec<OrderingTarget>,
+   strict_ordering_after: Vec<OrderingTarget>,
+   // The `TypeId` here is the type id of the flushing system
+   flushed_ordering_before: Vec<(TypeId, OrderingTarget)>,
+   flushed_ordering_after: Vec<(TypeId, OrderingTarget)>,
+   run_criteria: Vec<SystemId>,
+   // The combined access requirements of the system and its run criteria
+   joint_access: FilteredAccessSet<ComponentId>,
+   sets: Vec<Box dyn SystemLabel>,
+}
+
+struct SystemSetConfig {
+   system_config: SystemConfig,
+}

```

Comment:
In a previous design, there were some properties that could only be expressed meaningfully for a set. Namely "ambiguity sets" (this design sucks and we shouldn't port it) and atomic groups (super cool, but not getting in any time soon).

That said, we should probably cut it for now?

Reply:
I think so. Since atomicity is a separate RFC (#46) we can easily add it later if needed. Otherwise, it would be quite easy to forget about it and never delete it.